lib/vector-core/src/fanout.rs

```diff
@@ -63,165 +62,330 @@ impl Fanout {
+        if let Some(i) = self.sinks.iter().position(|(n, _)| n == id) {
+            let (_id, _removed) = self.sinks.remove(i);
+        }
+    }
 
-        if let Some(mut removed) = removed {
-            tokio::spawn(async move { removed.close().await });
+    /// Replace an existing sink as an output.
+    ///
+    /// If the `sink` passed is `None`, operation of the `Fanout` will be paused until a `Some`
+    /// with the same key is received. This allows for cases where the previous version of
+    /// a stateful sink must be dropped before the new version can be created.
+    fn replace(&mut self, id: ComponentKey, sink: Option<BufferSender<Event>>) {
+        if let Some((_, existing)) = self.sinks.iter_mut().find(|(n, _)| n == &id) {
+            *existing = sink;
+        } else {
+            // send might have been in-flight at the time, in which case it will be cancelled
+            // panic!("Tried to replace a sink that's not already present");
+            self.sinks.push((id, sink));
         }
+    }
 
-        if self.i > i {
-            self.i -= 1;
+    fn contains(&mut self, id: &ComponentKey) -> bool {
+        self.sinks.iter().any(|(n, _)| n == id)
+    }
+
+    /// Process any available control messages, without blocking to wait for more.
+    pub fn process_control_messages(&mut self) {
+        while let Ok(message) = self.control_channel.try_recv() {
+            self.handle_control_message(message);
         }
     }
 
-    fn replace(&mut self, id: &ComponentKey, sink: Option<GenericEventSink>) {
-        if let Some((_, existing)) = self.sinks.iter_mut().find(|(n, _)| n == id) {
-            *existing = sink;
-        } else {
-            panic!("Tried to replace a sink that's not already present");
+    fn handle_control_message(&mut self, message: ControlMessage) {
+        match message {
+            ControlMessage::Add(id, sink) => self.add(id, sink),
+            ControlMessage::Remove(id) => self.remove(&id),
+            ControlMessage::Replace(id, sink) => self.replace(id, sink),
         }
     }
 
-    pub fn process_control_messages(&mut self, cx: &mut Context<'_>) {
-        while let Poll::Ready(Some(message)) = self.control_channel.poll_recv(cx) {
-            match message {
-                ControlMessage::Add(id, sink) => self.add(id, sink),
-                ControlMessage::Remove(id) => self.remove(&id),
-                ControlMessage::Replace(id, sink) => self.replace(&id, sink),
+    /// If any sink is awaiting replacement (i.e. it was temporarily replaced with `None`), read
+    /// and process messages from the control channel until that is no longer true.
+    async fn wait_for_replacements(&mut self) {
+        while self.sinks.iter().any(|x| x.1.is_none()) {
+            if let Some(msg) = self.control_channel.recv().await {
+                self.handle_control_message(msg);
+            } else {
+                // control channel is closed? probably doesn't matter what we do
```

Comment:
The control channel only goes off when Vector is shutting down, yes? Or crashing.

Reply:
I believe so, yes. Really we should never see this.